src/main/java/io/pravega/connectors/flink/FlinkPravegaOutputFormat.java

```diff
@@ -0,0 +1,279 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.pravega.connectors.flink;
+
+import io.pravega.client.ClientConfig;
+import io.pravega.client.ClientFactory;
+import io.pravega.client.stream.EventStreamWriter;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.io.OutputFormat;
+import org.apache.flink.api.common.io.RichOutputFormat;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A Flink {@link OutputFormat} that can be added as a sink to write into Pravega.
+ */
+@Slf4j
+public class FlinkPravegaOutputFormat<T> extends RichOutputFormat<T> {
+
+    private static final long serialVersionUID = 1L;
+
+    // The name of Pravega stream to write into.
+    private final String stream;
+
+    // The name of Pravega scope where the stream belongs to.
+    private final String scope;
+
+    // The supplied event serializer.
+    private final SerializationSchema<T> serializationSchema;
+
+    // The factory used to create Pravega clients; closing this will also close all Pravega connections.
+    private transient ClientFactory clientFactory;
+
+    // The Pravega client config.
+    private final ClientConfig clientConfig;
+
+    // The router used to partition events within a stream.
+    private final PravegaEventRouter<T> eventRouter;
+
+    // Pravega event writer instance.
+    private transient EventStreamWriter<T> pravegaWriter;
+
+    // Error which will be detected asynchronously and reported to Flink.
+    private final AtomicReference<Throwable> writeError;
+
+    // Used to track confirmation from all writes to ensure guaranteed writes.
+    private final AtomicInteger pendingWritesCount;
+
+    private transient ExecutorService executorService;
+
+    /**
+     * Creates a new Flink Pravega {@link OutputFormat} which can be added as a sink to a Flink batch job.
+     *
+     * @param clientConfig          The Pravega client configuration.
+     * @param stream                The stream to write the events.
+     * @param serializationSchema   The implementation to serialize events that will be written to pravega stream.
+     * @param eventRouter           The event router to be used while writing the events.
+     */
+    protected FlinkPravegaOutputFormat(
```

Comment:
Is this specific to tables? It seems to be a generic output format implementation. I just need a clarification.

Reply:
It is a generic implementation but used in `FlinkPravegaTableSink`. I have updated the class comment to mention about this.